From 755e2a5ff7571360059b7b0b6dc72b4a4d034c25 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:11:21 -0400
Subject: [powerpc] powernv: Check sysparam size before creation

Message-id: <1410545655-205645-453-git-send-email-dzickus@redhat.com>
Patchwork-id: 94613
O-Subject: [RHEL7 PATCH 452/626] powerpc/powernv: Check sysparam size before creation
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 63aecfb20a6629f1ca91d7a052f87988ceb6dd53
Author: Joel Stanley <joel@jms.id.au>
Date:   Thu Apr 24 16:55:37 2014 +0930

    powerpc/powernv: Check sysparam size before creation

    The size of the sysparam sysfs files is determined from the device tree
    at boot. However the buffer is hard coded to 64 bytes. If we encounter a
    parameter that is larger than 64, or miss-parse the device tree, the
    buffer will overflow when reading or writing to the parameter.

    Check it at discovery time, and if the parameter is too large, do not
    create a sysfs entry for it.

    Signed-off-by: Joel Stanley <joel@jms.id.au>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/powernv/opal-sysparam.c b/arch/powerpc/platforms/powernv/opal-sysparam.c
index c59e77a..d202f9b 100644
--- a/arch/powerpc/platforms/powernv/opal-sysparam.c
+++ b/arch/powerpc/platforms/powernv/opal-sysparam.c
@@ -241,6 +241,12 @@ void __init opal_sys_param_init(void)
 
  /* For each of the parameters, populate the parameter attributes */
  for (i = 0; i < count; i++) {
+  if (size[i] > MAX_PARAM_DATA_LEN) {
+   pr_warn("SYSPARAM: Not creating parameter %d as size "
+    "exceeds buffer length\n", i);
+   continue;
+  }
+
   sysfs_attr_init(&attr[i].kobj_attr.attr);
   attr[i].param_id = id[i];
   attr[i].param_size = size[i];
-- 
1.7.1